Common Node API (#1027)
filecoin-project/FIPs: FIPS/fip-common_node_api.md

---
fip: "<to be assigned>"
title: Filecoin Common Node Interface V1
author: "David Ansermino (@ansermino), Aatif Syed (@aatifsyed)"
discussions-to: <URL>
status: Draft

> type: FTP

Reviewer: ```suggestion
type: FIP - Interface
```

## Specification
<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Filecoin implementations. -->

Using the OpenRPC specification format we define several groups of methods. These methods are defined in the OpenRPC specification found here: https://github.com/ChainSafe/filecoin-openrpc-spec. A list of the methods is included below for posterity.

All methods are reproduced identically to their present implementation in Lotus 1.26.1.

Categories have been assigned to indicate the identified use case (see Design Rationale).

Node implementers must include all specified methods in their API without any modifications. Implementers may choose to provide additional methods beyond what is included in this specification. The endpoint used to serve RPC requests must be versioned acccording the majo

## Subscription methods

> The `ChainNotify` method is not a simple request/response, but rather a bidirectional exchange of JSON-RPC-like messages as follows:

Reviewer 2: Note: F3 will also introduce some form of `FinalizedChainNotify()` method which should be significantly simpler...

We should also consider:

1. Events (which currently use a different API).
2. Messages (there's actually no great "native" API for watching for messages).

Reviewer 3: Thanks for your comment Steven.

We're currently rewriting this section if you'd like to sneak a preview: https://github.com/ChainSafe/FIPs/pull/2

Could you clarify what the `Events` and `Messages` APIs you're talking about are, or maybe point me to some docs? I've not heard of them before.

Reviewer: https://github.com/filecoin-project/lotus/blob/f8c5b737efa10b84c649ae2471030d691902e688/api/api_full.go#L197 This is basically Filecoin’s version of eth_getlog

Reviewer: Messages are the ones started with `Mpool`

1. Caller sends a JSON-RPC request to the endpoint, with no parameters.
    ```json
    { "jsonrpc": "2.0", "id": 1, "method": "Filecoin.ChainNotify" }
    ```
2. Callee sends a JSON-RPC response with an integer Channel ID
    ```json
    { "jsonrpc": "2.0", "id": 1, "result": 2}
    ```
3. Callee sends JSON-RPC notifications, with a pair of positional parameters. The first is the Channel ID, the second is the actual payload
    ```json

>     { "jsonrpc": "2.0", "method": "xrpc.ch.val", params: [2, {}] }
>     ```
> 4. Caller sends a JSON-RPC request to cancel the subscription, with their Channel ID
>     ```json
>     { "jsonrpc": "2.0", "id": 1, "method": "xrpc.cancel", params: [2] }
>     ```
> 5. Callee responds with a JSON-RPC notification of cancellation
>     ```json
>     { "jsonrpc": "2.0", "id": 1, "method": "xrpc.ch.close", params: [2] }

Reviewer 4: ```suggestion
    { "jsonrpc": "2.0", "method": "xrpc.ch.val", "params": [2, {}] }
    ```
4. Caller sends a JSON-RPC request to cancel the subscription, with their Channel ID
    ```json
    { "jsonrpc": "2.0", "id": 1, "method": "xrpc.cancel", "params": [2] }
    ```
5. Callee responds with a JSON-RPC notification of cancellation
    ```json
    { "jsonrpc": "2.0", "id": 1, "method": "xrpc.ch.close", "params": [2] }
```

Reviewer 3: Thanks! We're actually reewriting this section - you can peep https://github.com/ChainSafe/FIPs/pull/2 to see the current progress :)

This interface is contrived from the existing [Lotus RPC V1 interface](https://github.com/filecoin-project/lotus/blob/48287574d82c81ffc50f259134c6596add32a2f5/documentation/en/api-v1-unstable-methods.md). Due to the large number of methods presently supported, we have limited the scope of this specification to include methods with a well established use case. We have observed the requirements of Lotus-miner, Curio, Web3Mine and Boost to inform which methods are actively used. We have also considered insights from Glif based on the usage of their RPC service.

We target two groups of users: *end users* and *node operators*. These each have their own requirements for particular data and access control. While no attempt is made in this FIP to specify considerations  for these groups (eg. access control), these groups are used to justify the need for particular methods.

Methods included in the API fall into one of these categories:

- State Queries

>   - These primarily concern the chain state, actor state or network (p2p) state. They are read-only methods that are necessary to expose all elements of global state.

Reviewer 2: Right now we have a pretty _wide_ API here. Maybe a bit of a bikeshed, but I've always wanted to move some of the state-query logic into read-only methods on the actors themselves so state query operations simply become method implementations (simplifying client implementations).

Author: Great suggestion! 

Another thought we had is whether a generic state query method could be sufficient. Having a more direct relationship between what the actors return and the RPC responses would make this more feasible.

Using the OpenRPC specification format we define several groups of methods. These methods are defined in the OpenRPC specification found here: https://github.com/ChainSafe/filecoin-openrpc-spec. A list of the methods is included below for posterity.

All methods are reproduced identically to their present implementation in Lotus 1.26.1.
Note that Lotus hosts two different API's at HTTP paths [`/rpc/v0`](https://github.com/filecoin-project/lotus/blob/v1.28.0-rc5/documentation/en/api-v0-methods.md) and [`/rpc/v1`](https://github.com/filecoin-project/lotus/blob/v1.28.0-rc5/documentation/en/api-v1-unstable-methods.md),
with the latter being marked as _unstable_.
This document aims to specify the `v1` API.

Categories have been assigned to indicate the identified use case (see Design Rationale).

Node implementers must include all specified methods in their API without any modifications. Implementers may choose to provide additional methods beyond what is included in this specification. The endpoint used to serve RPC requests must be versioned acccording the majo

## Subscription methods

> Certain methods like `ChainNotify` are not oneshot request/response exchanges,

Reviewer 5: I have brought up my concern regarding such push-based APIs [here](https://github.com/filecoin-project/lotus/pull/12719#issuecomment-2513994235). 

Without concrete usecases, clear delivery semantics (both in terms of guarantees, and latency) I advise against introducing such push-based APIs in favour of a simpler more maintainable pull-based API.

My nudge is to research the usecases carefully before introducing this style of API, and do it now because v2 APIs are a good place to de-clutter the APIs exposed for a set that is more focused on correctness+simplicity vs. featurefulness.

I believe that goes a long way as far as builders are concerned. 

I am keen to hear from the users of existing push-based APIs.

Author: Thanks for raising this! 

I believe it was the SP software that was highly dependent on `ChainNotify` in particular. This perhaps also suggests that an API for SPs should be defined independently of the "end user" API, they should maybe even use different protocols 🤷 

I will move to remove these from the proposed spec. As implementations already have support for this it should have no immediate impact on SPs.

Reviewer 2: `ChainNotify` has reasonably well-defined semantics (although they're not perfect):

- When you subscribe, you get a single event telling you the current "head", followed by all changes from that head.
- If you read slowely, you'll be unsubscribed.
- If you need the "full sequence", you can use `ChainGetPath` from your last tipset to the current head to fill in any missing pieces.

I'd prefer it if `ChainNotify(ctx)` were changed to `ChainNotify(ctx, from /* optional */)`, handling catch-up internally. But having a way to react to events immediately without having to poll is really nice.

NOTE: polling v. subscribing was a huge issue for F3 but, IMO, that was because subscribing had side effects (where `ChainNotify` does not).

Reviewer 5: > where `ChainNotify` does not

Indeed. FWIW, even for F3 I don't think push would have made things any easier to get correct.

> having a way to react to events immediately without having to poll is really nice.

I have blindspots here; apologies. What is the concrete usecase that is common enough to grant having this push API in Common API specification?

| ChainGetTipSetAfterHeight           | State query      | Looks back for a tipset at the specified epoch. If there are no blocks at the specified epoch, the first non-nil tipset at a later epoch will be returned. A fork can be specified, or null for the heaviest chain                                                                                                                                                                                                                             |
| ChainGetTipSetByHeight              | State query      | Returns the tipset at the specified height                                                                                                                                                                                                                                                                                                                                                                                                     |

> | ChainHasObj                         | State query      | checks if a given CID exists in the chain blockstore                                                                                                                                                                                                                                                                                                                                                                                           |

Reviewer 4: ```suggestion
| ChainHasObj                         | Node operation   | Checks if a given CID exists in the chain blockstore                                                                                                                                                                                                                                                                                                                                                                                           |
```

Reviewer 4: Well, currently these operations read from _the_ blockstore, not some special segment of it that's for state, it'll read anything with a CID, messages, events, whatever. I'm mainly taking issue with it being "state" query. Not a big deal, just seems to be a slight misscategorisation. ("Chain" in the name is also slightly misleading, but less so than "state")

| ChainHead                           | State query      | Returns the chain head (ie. the heaviest tipset)                                                                                                                                                                                                                                                                                                                                                                                               |
| ChainNotify                         | State query      | Subscribe to changes to the chain head                                                                                                                                                                                                                                                                                                                                                                                                         |

> | ChainPutObj                         |                  | Puts a given object into the block store                                                                                                                                                                                                                                                                                                                                                                                                       |

Reviewer 4: ```suggestion
| ChainPutObj                         | Node operation   | Puts a given object into the block store                                                                                                                                                                                                                                                                                                                                                                                                       |
```

> | ChainReadObj                        | State query      | Reads IPLD nodes referenced by the specified CID from chain blockstore and returns raw bytes.                                                                                                                                                                                                                                                                                                                                                  |

Reviewer 4: ```suggestion
| ChainReadObj                        | Node operation   | Reads IPLD nodes referenced by the specified CID from chain blockstore and returns raw bytes.                                                                                                                                                                                                                                                                                                                                                  |
```

| StateMinerFaults                    | State query      | Returns a bitfield indicating the faulty sectors of the given miner                                                                                                                                                                                                                                                                                                                                                                            |
| StateMinerInfo                      | State query      | returns info about the specified miner                                                                                                                                                                                                                                                                                                                                                                                                         |

> | StateMinerInitialPledgeCollateral   | State query      | returns the initial pledge collateral for the specified miner's sector                                                                                                                                                                                                                                                                                                                                                                         |

Reviewer 4: FYI we're going to have to deprecate `StateMinerInitialPledgeCollateral` and introduce a replacement because we can no longer use precommit to perform this calculation, so this is mostly useless now. Currently the new one is called `StateMinerInitialPledgeForSector` in https://github.com/filecoin-project/lotus/pull/12384

Author: Thanks for the heads up! WIll look into this and update the included methods :+1:

This interface is contrived from the existing [Lotus RPC V1 interface](https://github.com/filecoin-project/lotus/blob/48287574d82c81ffc50f259134c6596add32a2f5/documentation/en/api-v1-unstable-methods.md). Due to the large number of methods presently supported, we have limited the scope of this specification to include methods with a well established use case. We have observed the requirements of Lotus-miner, Curio, Web3Mine and Boost to inform which methods are actively used. We have also considered insights from Glif based on the usage of their RPC service.

We target two groups of users: *end users* and *node operators*. These each have their own requirements for particular data and access control. While no attempt is made in this FIP to specify considerations  for these groups (eg. access control), these groups are used to justify the need for particular methods.

Methods included in the API fall into one of these categories:

- State Queries
  - These primarily concern the chain state, actor state or network (p2p) state. They are read-only methods that are necessary to expose all elements of global state.

- Miner Operations

>   - A small set of specialized methods miner applications require to participate in consensus.

Reviewer 4: yes, but arguably mpoolpush and wallet* operations are useful for "client" operations too; maybe a broader category name would be good here, or split it into two?

Reviewer 4: although, MpoolPushMessage is "miner" and the other mpool* are "transactions"

maybe all mpool and all wallet should be "transactions" or make a new "client" category to replace "transactions"?
